### PR TITLE
add MaskRCNN

### PR DESCRIPTION
**Pull request by yuyu2172**
~Merge after #778~

### CODE REVIEW

**Comment by yuyu2172** [comment]
Wierd segfault during training...

```
 *** Process received signal ***........]  1.42%                                                                        
 Signal: Segmentation fault (11)........] 17.32%                                                                        
 Signal code: Address not mapped (1)                                                                                    
 Failing at address: 0x67ee6000s, 21:36:26.539275.                                                                      
 [ 0] /lib/x86_64-linux-gnu/libpthread.so.0(+0x11390)[0x7fbb433f5390]                                                   
 [ 1] /usr/local/lib/python3.6/site-packages/cv2.cpython-36m-x86_64-linux-gnu.so(+0x7971a0)[0x7fb9d82c31a0]             
 [ 2] /usr/local/lib/python3.6/site-packages/cv2.cpython-36m-x86_64-linux-gnu.so(+0x61ce58)[0x7fb9d8148e58]             
 [ 3] /usr/local/lib/python3.6/site-packages/cv2.cpython-36m-x86_64-linux-gnu.so(+0x7cf482)[0x7fb9d82fb482]             
 [ 4] /usr/local/lib/python3.6/site-packages/cv2.cpython-36m-x86_64-linux-gnu.so(+0x7d0879)[0x7fb9d82fc879]             
 [ 5] /usr/local/lib/python3.6/site-packages/cv2.cpython-36m-x86_64-linux-gnu.so(+0x2ef784)[0x7fb9d7e1b784]             
 [ 6] /usr/local/lib/libpython3.6m.so.1.0(_PyCFunction_FastCallDict+0x13c)[0x7fbb43733d6c]                              
 [ 7] /usr/local/lib/libpython3.6m.so.1.0(_PyCFunction_FastCallKeywords+0x4d)[0x7fbb43733bfd]                           
 [ 8] /usr/local/lib/libpython3.6m.so.1.0(+0x18b9fb)[0x7fbb4378c9fb]                                                    
 [ 9] /usr/local/lib/libpython3.6m.so.1.0(_PyEval_EvalFrameDefault+0x1363)[0x7fbb43785a83]                              
 [10] /usr/local/lib/libpython3.6m.so.1.0(+0x18be0a)[0x7fbb4378ce0a]                                                    
 [11] /usr/local/lib/libpython3.6m.so.1.0(+0x18badb)[0x7fbb4378cadb]                                                    
 [12] /usr/local/lib/libpython3.6m.so.1.0(_PyEval_EvalFrameDefault+0x34f)[0x7fbb43784a6f]                               
 [13] /usr/local/lib/libpython3.6m.so.1.0(+0x18c0f3)[0x7fbb4378d0f3]                                                    
 [14] /usr/local/lib/libpython3.6m.so.1.0(+0x18badb)[0x7fbb4378cadb]                                                    
 [15] /usr/local/lib/libpython3.6m.so.1.0(_PyEval_EvalFrameDefault+0x1363)[0x7fbb43785a83]                              
 [16] /usr/local/lib/libpython3.6m.so.1.0(+0x18be0a)[0x7fbb4378ce0a]                                                    
 [17] /usr/local/lib/libpython3.6m.so.1.0(+0x18badb)[0x7fbb4378cadb]                                                    
 [18] /usr/local/lib/libpython3.6m.so.1.0(_PyEval_EvalFrameDefault+0x34f)[0x7fbb43784a6f]                               
 [19] /usr/local/lib/libpython3.6m.so.1.0(_PyFunction_FastCallDict+0x41b)[0x7fbb4378eacb]                               
 [20] /usr/local/lib/libpython3.6m.so.1.0(_PyObject_FastCallDict+0x10e)[0x7fbb436fa86e]                                 
 [21] /usr/local/lib/libpython3.6m.so.1.0(_PyObject_Call_Prepend+0x61)[0x7fbb436fb1f1]                                  
 [22] /usr/local/lib/libpython3.6m.so.1.0(PyObject_Call+0x47)[0x7fbb436fac17]                                           
 [23] /usr/local/lib/libpython3.6m.so.1.0(+0x147506)[0x7fbb43748506]                                                    
 [24] /usr/local/lib/libpython3.6m.so.1.0(PyObject_Call+0x47)[0x7fbb436fac17]                                           
 [25] /usr/local/lib/libpython3.6m.so.1.0(_PyEval_EvalFrameDefault+0x1c69)[0x7fbb43786389]                              
 [26] /usr/local/lib/libpython3.6m.so.1.0(_PyFunction_FastCallDict+0x41b)[0x7fbb4378eacb]                               
 [27] /usr/local/lib/libpython3.6m.so.1.0(_PyObject_FastCallDict+0x10e)[0x7fbb436fa86e]                                 
 [28] /usr/local/lib/libpython3.6m.so.1.0(_PyObject_Call_Prepend+0x61)[0x7fbb436fb1f1]                                  
 [29] /usr/local/lib/libpython3.6m.so.1.0(PyObject_Call+0x47)[0x7fbb436fac17]                                           
 *** End of error message ***                                                                                           
```

**Comment by yuyu2172** [comment]
```
$ mpiexec -n 8 python3 train_multi.py --batchsize 8
--------------------------------------------------------------------------
WARNING: There are more than one active ports on host 'coco1-worker-0', but the
default subnet GID prefix was detected on more than one of these
ports.  If these ports are connected to different physical IB
networks, this configuration will fail in Open MPI.  This version of
Open MPI requires that every physically separate IB subnet that is
used between connected MPI processes must have different subnet ID
values.

Please see this FAQ entry for more details:

  http://www.open-mpi.org/faq/?category=openfabrics#ofa-default-subnet-gid

NOTE: You can turn off this warning by setting the MCA parameter
      btl_openib_warn_default_gid_prefix to 0.
--------------------------------------------------------------------------
[coco1-worker-0:115997] 7 more processes have sent help message help-mpi-btl-openib.txt / default subnet prefix
[coco1-worker-0:115997] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
0 (1, 3, 1344, 576)
6 (1, 3, 800, 1216)
4 (1, 3, 1088, 800)
7 (1, 3, 800, 1216)
15 (1, 3, 800, 1088)
19 (1, 3, 1088, 800)
19 (1, 3, 704, 1344)
41 (1, 3, 800, 1088)
[coco1-worker-0:116005] Failed to cuMemcpy GPU memory, rc=-1
--------------------------------------------------------------------------
The call to cuMemcpyAsync failed. This is a unrecoverable error and will
cause the program to abort.
  cuMemcpyAsync(0x7f5f2f22ac00, 0x7f5df4dfb000, 8192) returned value 1
Check the cuda.h file for what the return value means.
--------------------------------------------------------------------------
Exception in main training loop: MPI_ERR_TRUNCATE: message truncated
Traceback (most recent call last):
  File "/home/yuyu2172/chainer/chainer/training/trainer.py", line 316, in run
    update()
  File "/home/yuyu2172/chainer/chainer/training/updaters/standard_updater.py", line 170, in update
    self.update_core()
  File "/home/yuyu2172/chainer/chainer/training/updaters/standard_updater.py", line 182, in update_core
    optimizer.update(loss_func, *in_arrays)
  File "/home/yuyu2172/chainer/chainermn/optimizers.py", line 28, in update
    self.communicator.bcast_data(target)
  File "/home/yuyu2172/chainer/chainermn/communicators/mpi_communicator_base.py", line 615, in bcast_data
    self.mpi_comm.Bcast(buf)
  File "mpi4py/MPI/Comm.pyx", line 579, in mpi4py.MPI.Comm.Bcast
Will finalize trainer extensions and updater before reraising the exception.
Traceback (most recent call last):
  File "train_multi.py", line 249, in <module>
    main()
  File "train_multi.py", line 245, in main
    trainer.run()
  File "/home/yuyu2172/chainer/chainer/training/trainer.py", line 349, in run
    six.reraise(*exc_info)
  File "/usr/local/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/home/yuyu2172/chainer/chainer/training/trainer.py", line 316, in run
    update()
  File "/home/yuyu2172/chainer/chainer/training/updaters/standard_updater.py", line 170, in update
    self.update_core()
  File "/home/yuyu2172/chainer/chainer/training/updaters/standard_updater.py", line 182, in update_core
    optimizer.update(loss_func, *in_arrays)
  File "/home/yuyu2172/chainer/chainermn/optimizers.py", line 28, in update
    self.communicator.bcast_data(target)
  File "/home/yuyu2172/chainer/chainermn/communicators/mpi_communicator_base.py", line 615, in bcast_data
    self.mpi_comm.Bcast(buf)
  File "mpi4py/MPI/Comm.pyx", line 579, in mpi4py.MPI.Comm.Bcast
mpi4py.MPI.Exception: MPI_ERR_TRUNCATE: message truncated
```

**Comment by yuyu2172** [comment]
Seemingly working with the current master
Chainer: https://github.com/chainer/chainer/commit/afe903389d822583a5355e9d46e6766d048ebeb5
CuPy: https://github.com/cupy/cupy/commit/155228fdf6bb148a4c5537dfe08fbfef5329e416

**Comment by yuyu2172** [comment]
For speeding up training process
### What I did
+ Use MultiprocessIterator
+ Do resize in different process created by MultiprocessIterator
+ Stop CPU/GPU copy of masks

### What I did not do
+ Use polygon representation of masks. Although this speeds up training as done by Detectron, I wanted the training script to work with instance segmentation datasets that do not have polygon representation of masks.

**Comment by yuyu2172** [comment]
The performance of the trained model at this moment (should be 33.9mmAP)
```
mmAP (all): 0.31496558    
mmAP (large): 0.46482706  
mmAP (medium): 0.33771247 
mmAP (small): 0.14785959  
```

**Comment by yuyu2172** [comment]
Finally reproduced. It was critical to implement`mask_to_segm` correctly. 2d44d66

```
mmAP (all): 0.34442434     
mmAP (large): 0.50526273   
mmAP (medium): 0.36985907  
mmAP (small): 0.16083415   
```

**Comment by yuyu2172** [comment]
Performance after deleting `+1 to bbox https://github.com/yuyu2172/chainercv/commit/6513e2480e98ce3773ec5566c9923be51f75bb23

```
mmAP (all): 0.33432382    
mmAP (large): 0.5024262   
mmAP (medium): 0.36352885 
mmAP (small): 0.14599578  
```

**Comment by yuyu2172** [comment]
After https://github.com/chainer/chainercv/pull/781/commits/7e1e3ecee5930c527c7f1c89cdd4719d826095e5 and https://github.com/chainer/chainercv/pull/798
```
mmAP (all): 0.33900946  
mmAP (large): 0.5053451 
mmAP (medium): 0.3659876
mmAP (small): 0.1571092 
```

**Comment by yuyu2172** [comment]
```
mmAP (all): 0.35994896
mmAP (large): 0.5378797
mmAP (medium): 0.39182574
mmAP (small): 0.16467538
```

**Comment by yuyu2172** [comment]
Uploaded Weight scores

ResNet50  Mask

```
mmAP (all): 0.34175715    
mmAP (large): 0.50817233  
mmAP (medium): 0.36931357 
mmAP (small): 0.15625913  
```

ResNet50 Bbox
```
mmAP (all): 0.37964782     
mmAP (large): 0.4980176    
mmAP (medium): 0.4137187   
mmAP (small): 0.22214551   
```

ResNet101 Mask
```
mmAP (all): 0.35996467     
mmAP (large): 0.53794646   
mmAP (medium): 0.39182332  
mmAP (small): 0.16468713   

```

ResNet101 Bbox
```
mmAP (all): 0.40394193    
mmAP (large): 0.5254983   
mmAP (medium): 0.4450884  
mmAP (small): 0.23563214  
```

**Comment by yuyu2172** [comment]
pfnCI, test this please

**Comment by yuyu2172** [comment]
pfnCI, test this please

**Comment by yuyu2172** [comment]
@Hakuyume 
Please review

**Comment by yuyu2172** [comment]
TODO: 
`mask_loss` -> `mask_head_loss`
`bbox_loss` -> `bbox_head_loss`

**Comment by yuyu2172** [comment]
pfnCI, test this please

**Comment by yuyu2172** [comment]
pfnCI, test this please

File: examples/fpn/README.md [hunk 1]
```diff
@@ -1,31 +1,37 @@
...
+# Examples of Faster R-CNN and related models using FPN [1, 2]
 
 ## Performance
 MS COCO 2017 Val
 
-| Model | Original | ChainerCV |
-|:-:|:-:|:-:|
-| Faster-RCNN FPN ResNet50 | 36.7 % [2] | 37.1 % |
-| Faster-RCNN FPN ResNet101 | 39.4 % [2] | 39.5 % |
+| Backbone | Supervision | Detectron (Bbox) | Ours (Bbox) | Detectron (Mask) | Ours (Mask)|
```
Comment (Hakuyume) [inline examples/fpn/README.md]: I'd like to unify the format of tables that compare our implementations with the original implementations. Currently, we are using a variety names of column, e.g. `Original/Reference/<the name of original framework>` and `ChainerCV/Ours`.

File: chainercv/links/model/fpn/faster_rcnn_fpn_resnet.py [hunk 2]
```diff
@@ -87,44 +104,52 @@ class FasterRCNNFPNResNet50(FasterRCNNFPNResNet):
...
+    Please refer to :class:`~chainercv.links.model.fpn.FasterRCNNFPNResNet`.
+
+    """
+
+    _base = ResNet50
+    _models = {
+        'coco': {
+            'param': {'n_fg_class': 80},
+            'url': 'https://chainercv-models.preferred.jp/'
+            'faster_rcnn_fpn_resnet50_mask_coco_trained_2019_03_15.npz',
```
Comment (Hakuyume) [inline chainercv/links/model/fpn/faster_rcnn_fpn_resnet.py]: How about `mask_rcnn_fpn_resnet50_coco_...`?

**Comment by yuyu2172** [comment]
pfnCI, test this please

File: examples/fpn/README.md (hunk 1 above)
Comment (yuyu2172) [reply examples/fpn/README.md]: `Original` is sometimes easier to use than the name of the project. For instance, `SegNet-Tutorial` is the name for SegNet. IMO, this is verbose.
How about unifying it to `Original` and `Ours`.

**Comment by Hakuyume** [comment]
pfnCI, test this please

File: examples/fpn/README.md (hunk 1 above)
Comment (Hakuyume) [reply examples/fpn/README.md]: > How about unifying it to Original and Ours.
OK. Could you update the table of FPN? The other tables should be fixed in another PR.

Comment (yuyu2172) [reply examples/fpn/README.md]: OK

File: chainercv/links/model/fpn/faster_rcnn_fpn_resnet.py (hunk 2 above)
Comment (yuyu2172) [reply chainercv/links/model/fpn/faster_rcnn_fpn_resnet.py]: Done

**Comment by yuyu2172** [comment]
pfnCI, test this please

**Comment by yuyu2172** [comment]
pfnCI, test this please

File: chainercv/links/model/fpn/faster_rcnn.py [hunk 3]
```diff
@@ -45,18 +48,36 @@ class FasterRCNN(chainer.Chain):
 
     """
 
-    _stride = 32
+    stride = 32
+    _accepted_return_values = ('rois', 'bboxes', 'labels', 'scores', 'masks')
```
Comment (Hakuyume) [inline chainercv/links/model/fpn/faster_rcnn.py]: `_acceptable_return_values`?

File: chainercv/links/model/fpn/faster_rcnn.py [hunk 4]
```diff
@@ -45,18 +48,36 @@ class FasterRCNN(chainer.Chain):
 
     """
 
-    _stride = 32
+    stride = 32
```
Comment (Hakuyume) [inline chainercv/links/model/fpn/faster_rcnn.py]: Why did you make this member public?

File: chainercv/links/model/fpn/faster_rcnn.py [hunk 5]
```diff
@@ -45,18 +48,36 @@ class FasterRCNN(chainer.Chain):
...
-            self.head = head
+            if self._run_bbox:
+                self.bbox_head = bbox_head
+            if self._run_mask:
+                self.mask_head = mask_head
 
-        self._min_size = min_size
-        self._max_size = max_size
+        self.min_size = min_size
+        self.max_size = max_size
```
Comment (Hakuyume) [inline chainercv/links/model/fpn/faster_rcnn.py]: Why did you make these members public?

File: chainercv/links/model/fpn/faster_rcnn.py [hunk 6]
```diff
@@ -45,18 +48,36 @@ class FasterRCNN(chainer.Chain):
...
+            if value_name not in self._accepted_return_values:
+                raise ValueError(
+                    '{} is not included in accepted value names {}'.format(
+                        value_name, self._accepted_return_values))
+        self._return_values = return_values
+
+        self._store_rpn_outputs = 'rois' in self._return_values
+        self._run_bbox = any([key in self._return_values
+                              for key in
+                              ['bboxes', 'labels', 'scores', 'masks']])
```
Comment (Hakuyume) [inline chainercv/links/model/fpn/faster_rcnn.py]: You can write as follows.
```python
self._run_bbox = bool(set(self._return_values) & {'bboxes', 'labels', 'scores', 'masks'})
```

File: chainercv/links/model/fpn/faster_rcnn.py [hunk 7]
```diff
@@ -94,52 +115,106 @@ def forward(self, x):
...
-               Each value indicates how confident the prediction is.
+            tuple of lists:
+            The table below shows the input and possible outputs.
+
+        .. csv-table::
+            :header: name, shape, dtype, format
+
+            :obj:`imgs`, ":math:`[(3, H, W)]`", :obj:`float32`, \
+            "RGB, :math:`[0, 255]`"
+            :obj:`rois`, ":math:`[(R', 4)]`", :obj:`float32`, \
```
Comment (Hakuyume) [inline chainercv/links/model/fpn/faster_rcnn.py]: Can we add a line between inputs part and outputs part? (or just make two tables)

File: chainercv/links/model/fpn/mask_utils.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+    H, W = size
+    _, segm_size, _ = segm.shape
+
+    mask = np.zeros((len(bbox), H, W), dtype=np.bool)
+
+    # To work around an issue with cv2.resize (it seems to automatically
+    # pad with repeated border values), we manually zero-pad the masks by 1
+    # pixel prior to resizing back to the original image resolution.
+    # This prevents "top hat" artifacts. We therefore need to expand
+    # the reference boxes by an appropriate factor.
```
Comment (Hakuyume) [inline chainercv/links/model/fpn/mask_utils.py]: We have two similar blocks. Can we merge them?

File: chainercv/links/model/fpn/faster_rcnn.py (hunk 4 above)
Comment (yuyu2172) [reply chainercv/links/model/fpn/faster_rcnn.py]: This value is necessary for Transform.

File: chainercv/links/model/fpn/faster_rcnn.py (hunk 5 above)
Comment (yuyu2172) [reply chainercv/links/model/fpn/faster_rcnn.py]: https://github.com/chainer/chainercv/pull/781#discussion_r276521963

File: chainercv/links/model/fpn/mask_head.py [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+        n_class (int): The number of classes including background.
+        scales (tuple of floats): The scales of feature maps.
+
+    """
+
+    _canonical_level = 2
+    _canonical_scale = 224
+    _roi_size = 14
+    _roi_sample_ratio = 2
+    segm_size = _roi_size * 2
```
[the remark this answers is not in the document]
Comment (yuyu2172) [reply chainercv/links/model/fpn/mask_head.py]: https://github.com/chainer/chainercv/pull/781#discussion_r276521963

File: chainercv/links/model/fpn/faster_rcnn.py (hunk 4 above)
Comment (yuyu2172) [reply chainercv/links/model/fpn/faster_rcnn.py]: It was not possible to reuse method of `FasterRCNN` inside `Transform`.
I had to expose these attributes.

File: chainercv/links/model/fpn/faster_rcnn.py (hunk 3 above)
Comment (yuyu2172) [reply chainercv/links/model/fpn/faster_rcnn.py]: OK

File: chainercv/links/model/fpn/faster_rcnn.py (hunk 6 above)
Comment (yuyu2172) [reply chainercv/links/model/fpn/faster_rcnn.py]: Thanks

File: chainercv/links/model/fpn/faster_rcnn.py (hunk 7 above)
Comment (yuyu2172) [reply chainercv/links/model/fpn/faster_rcnn.py]: OK

File: chainercv/links/model/fpn/mask_utils.py (hunk 8 above)
Comment (yuyu2172) [reply chainercv/links/model/fpn/mask_utils.py]: OK

File: chainercv/links/model/fpn/mask_head.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Hakuyume) [reply chainercv/links/model/fpn/mask_head.py]: I see. It is ok to make this member public.

**Comment by yuyu2172** [comment]
1. Define `min_size` and `max_size` in `train_multi.py`.
2. use only cv2 backend in `mask_utils.py`
   + Change the behavior of `resize`. Make it break when `chainer.config.cv_resize_backend == 'cv2'` and `cv2` is not installed.
   + Force `cv2` backend in `mask_utils.py`

**Comment by yuyu2172** [comment]
pfnCI, test this please

**Comment by Hakuyume** [comment]
pfnCI, test this please

**Comment by yuyu2172** [comment]
pfnCI, test this please

**Comment by yuyu2172** [comment]
pfnCI, test this please

**Pull request merged**